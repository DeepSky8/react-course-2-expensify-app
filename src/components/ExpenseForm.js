import React from 'react';
import moment from 'moment';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt).format("yyyy-MM-DD") : moment().format("yyyy-MM-DD"),
            calendarFocused: false,
            error: props.error ? props.error : ''
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }))
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = (e) => {
        const createdAt = e.target.value;
        this.setState(() => ({ createdAt }));

    }
    // onFocusChange = ({ focused }) => {

    //     this.setState(() => { calendarFocused: focused })
    // }
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount' }))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <label htmlFor='description'>Description
                        <input
                            id='description'
                            name='description'
                            type="text"
                            placeholder="Required"
                            autoFocus
                            value={this.state.description}
                            onChange={(this.onDescriptionChange)}
                        />
                    </label>
                    <label htmlFor='amount'>Amount
                        <input
                            id='amount'
                            name='amount'
                            type="text"
                            placeholder="Required in numbers"
                            value={this.state.amount}
                            onChange={this.onAmountChange}
                        />
                    </label>
                    <label htmlFor='date'>Date
                        <input
                            id='date'
                            name='date'
                            type="date"
                            value={this.state.createdAt}
                            onChange={this.onDateChange}
                        />
                    </label>
                    <label htmlFor='note'>Note (optional)
                        <textarea
                            id='note'
                            name='note'
                            placeholder="Add a note for your expense (optional)"
                            value={this.state.note}
                            onChange={this.onNoteChange}
                        >
                        </textarea>
                    </label>
                    <button>Save Expense</button>
                </form>
            </div>
        )
    }
}
