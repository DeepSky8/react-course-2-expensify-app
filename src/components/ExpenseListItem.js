import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
    <div>
        <nav>
            <Link to={`edit/${id}`}>
                <h3>{description}</h3>
            </Link>
        </nav>

        <p>Amount: {amount}</p>
        <p>Created: {createdAt}</p>
    </div>
)


export default ExpenseListItem;