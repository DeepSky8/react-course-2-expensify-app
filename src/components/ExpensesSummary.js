import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses'
import numeral from 'numeral';
import expensesSummaryData from '../selectors/expenses-summaryData';

const ExpensesSummary = (props) => {
    const summaryData = expensesSummaryData(props.expenses);
    const arrayLength = summaryData.arrayLength.toString();
    const arrayTotal = numeral(summaryData.arrayTotal / 100).format('$0,0.00');
    const plural = arrayLength === 1 ? '' : 's';

    return (
        <div>
            <p>
                {
                    `Viewing ${arrayLength} 
                expense${plural} 
                totalling ${arrayTotal}`
                }
            </p>

        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(ExpensesSummary)