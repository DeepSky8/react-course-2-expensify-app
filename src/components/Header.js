import React from 'react';
import { NavLink } from 'react-router-dom';
import { startLogout } from '../actions/auth';
import { connect } from 'react-redux';



export const Header = ({ startLogout }) => (
    <header>
        <h1>Expensify</h1>
        <p>
            <NavLink to="/dashboard" className={({ isActive }) =>
                isActive ? 'is-active' : undefined
            }>
                Dashboard
            </NavLink>
            <NavLink to="/create" className={({ isActive }) =>
                isActive ? 'is-active' : undefined
            }>
                Create Expense
            </NavLink>
            <button name='logout' onClick={startLogout}>Logout</button>
        </p>
    </header>
)

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
})

export default connect(undefined, mapDispatchToProps)(Header);

// <NavLink to="/help" className={({ isActive }) =>
//                 isActive ? 'is-active' : undefined
//             }>
//                 Help
//             </NavLink>

// <NavLink to="/" className={({ isActive }) =>
// isActive ? 'is-active' : undefined
// }>Login
// </NavLink>