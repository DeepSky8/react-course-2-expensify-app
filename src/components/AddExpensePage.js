import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';
import { useNavigate } from 'react-router-dom';
import { history } from '../routers/AppRouter';





export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {

        this.props.startAddExpense(expense);
        history.push('/')
    }
    render() {
        return (
            <div>
                <div className='page-header'>
                    <div className='content-container'>
                        <h1 className='page-header__title'>Add Expense</h1>
                    </div>
                </div>
                <div className='content-container'>
                    <ExpenseForm
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        )
    }
}


const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);



// const AddExpensePage = (props) => {
//     const navigate = useNavigate();
//     return (
//         <div>
//             <h1>Add Expense</h1>
//             <ExpenseForm
//                 onSubmit={(expense) => {
//                     props.dispatch(startAddExpense(expense))
//                     navigate('/dashboard')
//                 }}
//             />
//         </div>
//     )
// }

// export default connect()(AddExpensePage);