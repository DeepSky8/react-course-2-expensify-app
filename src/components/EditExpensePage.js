import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm';
import { useNavigate } from 'react-router-dom';



const EditExpensePage = (props) => {
    const navigate = useNavigate();
    return (
        <div>
            <ExpenseForm
                expense={props.expense}
                onSubmit={(expense) => {
                    props.dispatch(startEditExpense(props.expense.id, expense))
                    navigate('/')
                }} />
            <button onClick={() => {
                props.dispatch(startRemoveExpense({ id: props.expense.id }))
                navigate('/')
            }}>Remove Expense</button>
        </div>
    )
}

const mapStateToProps = (state) => {
    const params = { id: window.location.pathname.split("/")[2] };
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === params.id
        })
    }
}

export default connect(mapStateToProps)(EditExpensePage);

