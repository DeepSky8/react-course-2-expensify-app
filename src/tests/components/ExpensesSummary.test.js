import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should display number and total of expenses to be 3, 1141.95', () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(<ExpensesSummary expenseCount={3} expensesTotal={114195} />)
    expect(renderer.getRenderOutput()).toMatchSnapshot();
 })

 test('should display number and total of expenses to be 0, 0', () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(<ExpensesSummary expenseCount={0} expensesTotal={0} />)
    expect(renderer.getRenderOutput()).toMatchSnapshot();
 })

 test('should display number and total of expenses to be 1, 1.95', () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(<ExpensesSummary expenseCount={1} expensesTotal={195} />)
    expect(renderer.getRenderOutput()).toMatchSnapshot();
 })