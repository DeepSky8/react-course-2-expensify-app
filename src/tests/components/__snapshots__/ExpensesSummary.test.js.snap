// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display number and total of expenses to be 0, 0 1`] = `
<div>
  <p>
    Viewing 0 expenses totalling $0.00
  </p>
</div>
`;

exports[`should display number and total of expenses to be 1, 1.95 1`] = `
<div>
  <p>
    Viewing 1 expense totalling $1.95
  </p>
</div>
`;

exports[`should display number and total of expenses to be 3, 1141.95 1`] = `
<div>
  <p>
    Viewing 3 expenses totalling $1,141.95
  </p>
</div>
`;
